### ISSUE TITLE
[AI][Ascend] Fix boolean-mask semantics in aten::index.Tensor

### ISSUE THREAD

**Issue by yswang777** [flagos-ai/Torch-FL#323]
## Issue Type
- [x] Bug Report
- [ ] Feature Request
- [x] Operator Implementation
- [ ] Platform Support
- [ ] Performance Issue
- [ ] Documentation

## AI Agent Information
- **Agent**: OpenAI Codex
- **Model**: GPT-5
- **Session Context**: Qwen-Image-2512 bring-up on Torch-FL/Ascend 910C exposed an incorrect `aten::index.Tensor` result for a boolean sequence mask. The model workaround was isolated and the native Ascend implementation was reviewed before filing.

## Summary

The Ascend `aten::index.Tensor` kernel treats boolean tensors as integer-style advanced indices when it computes the broadcast and output shapes. A boolean prefix mask over a `[B, S, H]` tensor therefore returns the wrong shape instead of consuming the masked dimensions and compacting the selected rows. Qwen-Image-2512 reaches this through `hidden_states[attention_mask]` in its text-encoding path.

## Environment
<details>
<summary>Click to expand environment details</summary>

- **Platform**: Ascend
- **Python**: 3.11.15
- **PyTorch**: 2.10.0+cpu
- **torch_fl**: `fe8fe3d5d6674ef357fae03810d11c9a59635b2d` (bug observed); current main checked at `90a184f06d5792d9cbdeeaf08e19e8eb813fe574`
- **Hardware**: Ascend 910C, 16 visible devices

**Build config:**
```bash
ACCELERATOR=ascend FLAGGEMS_KERNEL=0 FLAGGEMS_PYTHON=1 \
CUDA_KERNEL=0 ASCEND_KERNEL=1 pip install --no-build-isolation -v -e .
```

**Runtime config:**
```bash
ACCELERATOR=ascend FLAGOS_DISABLE_CUDA_ASSETS=1 python repro.py
```
</details>

## Reproduction
```python
import torch_fl
import torch

x_cpu = torch.arange(15, dtype=torch.float32).reshape(1, 5, 3)
mask_cpu = torch.tensor([[True, True, False, True, False]])

expected = x_cpu[mask_cpu]
actual = x_cpu.to("flagos:0")[mask_cpu.to("flagos:0")].cpu()

print("expected", expected.shape, expected.tolist())
print("actual", actual.shape, actual.tolist())
torch.testing.assert_close(actual, expected)
```

## Expected vs Actual Behavior
**Expected:**

The mask consumes the leading two dimensions and returns shape `(3, 3)` with the three selected rows.

**Actual:**

The native Ascend kernel derives its output from the mask's broadcast shape and produces an incorrect rank/shape (observed as `(1, 5, 3)` in the real probe), so the text encoder cannot use the stock Diffusers path.

## Root Cause Analysis

`IndexTensorKernelAscend` collects every defined index tensor into one list, computes their ordinary broadcast shape, and uses that shape as the replacement for one indexed dimension. That is valid for integer tensor indices but not for boolean masks. PyTorch boolean indexing consumes `mask.dim()` source dimensions and produces a data-dependent leading dimension equal to the number of true elements. The implementation never branches on `kBool`, never counts selected elements, and allocates the output using integer-index rules before calling `aclnnIndex`.

## Proposed Solution

Add an explicit boolean-mask path to the Ascend index implementation. It must implement PyTorch prefix-mask semantics, including the number of consumed source dimensions and the data-dependent selected-row count, while leaving the existing integer advanced-index path unchanged. The existing Ascend `masked_select` implementation demonstrates the accepted pattern for a data-dependent output length, but the index result must preserve trailing dimensions rather than flattening them.

Do not solve this by patching Diffusers to turn a right-padded mask into a slice; that is a model-specific workaround and would leave `aten::index.Tensor` incorrect.

## Verification Plan
- [ ] Add Ascend integration tests for single 1-D and 2-D boolean prefix masks.
- [ ] Cover all-true, all-false, sparse, empty, and non-contiguous masks.
- [ ] Cover trailing feature dimensions and compare shape, dtype, device, and values with CPU.
- [ ] Verify integer tensor indexing remains unchanged.
- [ ] Run the Qwen-Image text-encoder stage without the mask-extraction monkeypatch.

## Context & Investigation

1. Reproduced the shape mismatch with the standalone script above and the Qwen attention-mask shape.
2. Read `csrc/aten/backends/ascend/index.cc` in full and compared its output-shape calculation with PyTorch boolean-index semantics.
3. Confirmed `torch_fl/configs/backends_ascend.conf` routes `index.Tensor` to `ascend`, so this is the selected native implementation rather than a CPU fallback.
4. Checked open issues and pull requests; no existing Ascend boolean-index fix was found. PR #87 concerns CUDA boxing of CPU index tensors and is a different path.

## Related Code Locations
- `csrc/aten/backends/ascend/index.cc:12` - native Ascend `IndexTensorKernelAscend`.
- `csrc/aten/backends/ascend/index.cc:29` - integer-style index broadcast calculation.
- `csrc/aten/backends/ascend/index.cc:49` - output-shape calculation that does not handle boolean masks.
- `csrc/aten/backends/ascend/masked_select.cc:19` - existing data-dependent Ascend output pattern.
- `torch_fl/configs/backends_ascend.conf:1160` - `index.Tensor = ascend`.

## Checklist - AI Agents MUST Complete All
- [x] I have provided complete environment information
- [x] I have included a minimal, self-contained reproducer
- [x] I have included the observed failure signature
- [x] I have analyzed the root cause (not just symptoms)
- [x] I have proposed a specific solution with implementation approach
- [x] I have identified affected code locations with line numbers
- [x] I have described how to verify the fix
- [x] I have checked for duplicate issues
- [x] All text is in **English**
- [x] Code follows project conventions (checked the existing Ascend handwritten kernels)
